### ISSUE TITLE
Setting default_available_locales to empty array breaks FastGettext.default_locale

### ISSUE THREAD

**Issue by joshcooper** [puppetlabs/gettext-setup-gem#61]
[`FastGettext.default_available_locales` defaults to nil](https://github.com/grosser/fast_gettext/blob/master/lib/fast_gettext/storage.rb#L36), which effectively disables the "available locales" machinery in fast gettext. Gettext-setup sets it to an [empty array](https://github.com/puppetlabs/gettext-setup-gem/blob/master/lib/gettext-setup/gettext_setup.rb#L14), so fast gettext [rejects all candidate locales](https://github.com/grosser/fast_gettext/blob/d4092023984badf91596f81ce81a3c7405ac003e/lib/fast_gettext/storage.rb#L130-L137), causing [`FastGettext.default_locale` to return nil](https://github.com/grosser/fast_gettext/blob/d4092023984badf91596f81ce81a3c7405ac003e/lib/fast_gettext/storage.rb#L118):

```ruby
require 'gettext-setup'

GettextSetup.initialize
puts "Default locale is #{FastGettext.default_locale.inspect}"
```

Produces:

```
$ bundle exec ruby locale.rb
Default locale is nil
```

I think we want to remove the line that sets it to an empty array.

/cc @Magisus

**Comment by joshcooper** [comment]
I ran into this while researching https://tickets.puppetlabs.com/browse/PUP-8407.

**Comment by smcclellan** [comment]
Nixing that line alone causes several other errors when we try to apply the default locale on line 40. That line could be changed to `FastGettext.default_available_locales ||= []`, which might do the trick?

**Comment by Magisus** [comment]
We can't use the `||` construct, because that will _ensure_ it is never nil, but we _want_ it to be nil.

I think it might be sufficient for our use case to move that line inside the initialize method. Since `initialize` no longer needs to be called prior to using the rake tasks, if we can stop calling `initialize` in the various rake files, and move that line, I think this bug will go away. However I'm not sure what impact that might have for other projects. If their use is consistent with what I've seen, it should be okay.

**Comment by joshcooper** [comment]
Hey @smcclellan can you take a look at this? If `gettext-setup` is required in the same ruby process as puppet, then it breaks all module CI. See https://tickets.puppetlabs.com/browse/PUP-8418. What @Magisus suggested makes sense to me:

```ruby
 def self.initialize(locales_path = 'locales', options = {})
   FastGettext.default_available_locales = nil
```

Also we need to think about how puppet and gettext-setup can safely interact, for example, semantic-puppet calls gettext-setup directly.

**Comment by smcclellan** [comment]
Just had a thought... is your config.yaml including a `default_locale`? The `initialize` method runs this: `FastGettext.default_locale = default_locale`

```
  def self.default_locale
    config['default_locale'] || 'en'
  end
```
The problem with making all the changes being suggested is that `FastGettext.default_available_locales` is `TrueClass`, which fails inside gettext and is the original reason for setting it to an array. I'm also debating adding this, to catch that specific case:
```
    FastGettext.default_available_locales = [] if FastGettext.default_available_locales == true
```

**Comment by Magisus** [comment]
If setting it to `nil` is causing the problem, I also don't see an issue with us setting it to `[]`, so long as it happens inside the `initialize` method and not when the class is required. What I'm trying to avoid here is having this gem manipulate the global internal state of FastGettext _on require_, because that means that for a lot of projects, just running `rake` causes that state to be altered (see what we had to do to avoid this in Puppet https://github.com/puppetlabs/puppet/commit/4ad1f430c09df91238e9710fb617786b36a08efa). We should only be calling `initialize` when we plan to actually use this gem at runtime to load translations, in which case we have a whole host of conflicts with Puppet's use of FastGettext, that should be resolved separately.

**Comment by smcclellan** [comment]
PR to move that is here: https://github.com/puppetlabs/gettext-setup-gem/pull/62

**Comment by smcclellan** [comment]
Fixed in gem version 0.30.
